File: desktop/src/shared/nativeCollectionLinks.ts
const INVALID_LINK = "INVALID_NATIVE_COLLECTION_LINK";
const INVALID_SCOPE = "INVALID_NATIVE_LINK_SCOPE";
const PLATFORM_SET = new Set(["XIAOHONGSHU", "DOUYIN", "BILIBILI", "ZHIHU"]);
const ROUTE_HOSTS = new Set([
  "www.bilibili.com",
  "space.bilibili.com",
  "www.douyin.com",
  "www.xiaohongshu.com",
  "www.zhihu.com",
  "zhuanlan.zhihu.com",
]);
const FORBIDDEN_DECODED = /[\s\\\p{Cc}\p{Cf}\p{Cs}\p{Zl}\p{Zp}]/u;
const QUERY_KEY = /^[A-Za-z0-9_.-]+$/;
const SENSITIVE_KEY = /(token|cookie|session|authorization|signature|password|secret)/;
const XSEC_TOKEN = /^[A-Za-z0-9_+\/=-]{1,1024}$/;
const XSEC_SOURCE = /^[A-Za-z0-9_-]{1,80}$/;

export type NativeCollectionLink = {
  platform: "XIAOHONGSHU" | "DOUYIN" | "BILIBILI" | "ZHIHU";
  kind: "detail" | "creator";
  external_id: string;
  canonical_url: string;
};

function invalidLink(): never {
  throw new Error(INVALID_LINK);
}

function decodeQueryPart(value: string): string {
  try {
    return decodeURIComponent(value.replace(/\+/g, "%20"));
  } catch {
    return invalidLink();
  }
}

function parseQuery(raw: string): Map<string, string> {
  const result = new Map<string, string>();
  if (raw === "") return result;
  for (const pair of raw.split("&")) {
    const separator = pair.indexOf("=");
    if (separator < 1) invalidLink();
    const key = decodeQueryPart(pair.slice(0, separator));
    const value = decodeQueryPart(pair.slice(separator + 1));
    if (!QUERY_KEY.test(key) || FORBIDDEN_DECODED.test(value)) invalidLink();
    const normalized = key.toLowerCase();
    if ((normalized === "xsec_token" || normalized === "xsec_source") && key !== normalized)
      invalidLink();
    if (result.has(normalized) || normalized === "modal_id") invalidLink();
    result.set(normalized, value);
  }
  return result;
}

function route(host: string, path: string): Omit<NativeCollectionLink, "canonical_url"> & { base: string } {
  let match: RegExpExecArray | null;
  if (host === "www.bilibili.com" && (match = /^\/video\/(BV[A-Za-z0-9]{10})\/?$/.exec(path)))
    return { platform: "BILIBILI", kind: "detail", external_id: match[1], base: `https://www.bilibili.com/video/${match[1]}` };
  if (host === "space.bilibili.com" && (match = /^\/([1-9][0-9]{0,19})\/?$/.exec(path)))
    return { platform: "BILIBILI", kind: "creator", external_id: match[1], base: `https://space.bilibili.com/${match[1]}` };
  if (host === "www.douyin.com" && (match = /^\/video\/([1-9][0-9]{0,19})\/?$/.exec(path)))
    return { platform: "DOUYIN", kind: "detail", external_id: match[1], base: `https://www.douyin.com/video/${match[1]}` };
  if (host === "www.douyin.com" && (match = /^\/user\/([A-Za-z0-9_-]{1,128})\/?$/.exec(path)))
    return { platform: "DOUYIN", kind: "creator", external_id: match[1], base: `https://www.douyin.com/user/${match[1]}` };
  if (host === "www.xiaohongshu.com" && (match = /^\/explore\/([0-9a-f]{24})\/?$/.exec(path)))
    return { platform: "XIAOHONGSHU", kind: "detail", external_id: match[1], base: `https://www.xiaohongshu.com/explore/${match[1]}` };
  if (host === "www.xiaohongshu.com" && (match = /^\/user\/profile\/([0-9a-f]{24})\/?$/.exec(path)))
    return { platform: "XIAOHONGSHU", kind: "creator", external_id: match[1], base: `https://www.xiaohongshu.com/user/profile/${match[1]}` };
  if (host === "www.zhihu.com" && (match = /^\/question\/([1-9][0-9]{0,19})\/answer\/([1-9][0-9]{0,19})\/?$/.exec(path)))
    return { platform: "ZHIHU", kind: "detail", external_id: `answer:${match[2]}`, base: `https://www.zhihu.com/question/${match[1]}/answer/${match[2]}` };
  if ((host === "www.zhihu.com" || host === "zhuanlan.zhihu.com") && (match = /^\/p\/([1-9][0-9]{0,19})\/?$/.exec(path)))
    return { platform: "ZHIHU", kind: "detail", external_id: `article:${match[1]}`, base: `https://zhuanlan.zhihu.com/p/${match[1]}` };
  if (host === "www.zhihu.com" && (match = /^\/zvideo\/([1-9][0-9]{0,19})\/?$/.exec(path)))
    return { platform: "ZHIHU", kind: "detail", external_id: `zvideo:${match[1]}`, base: `https://www.zhihu.com/zvideo/${match[1]}` };
  if (host === "www.zhihu.com" && (match = /^\/people\/([A-Za-z0-9_-]{1,128})\/?$/.exec(path)))
    return { platform: "ZHIHU", kind: "creator", external_id: match[1], base: `https://www.zhihu.com/people/${match[1]}` };
  return invalidLink();
}

export function parseNativeCollectionLink(value: unknown): NativeCollectionLink {
  if (typeof value !== "string" || value.length < 1 || value.length > 2048 || !/^[\x21-\x7e]+$/.test(value) || value.includes("\\") || value.includes("#"))
    return invalidLink();
  if (!value.startsWith("https://")) return invalidLink();
  const authorityEnd = value.indexOf("/", "https://".length);
  if (authorityEnd < 0) return invalidLink();
  const rawAuthority = value.slice("https://".length, authorityEnd);
  if (!ROUTE_HOSTS.has(rawAuthority)) return invalidLink();
  const queryStart = value.indexOf("?", authorityEnd);
  const rawPath = value.slice(authorityEnd, queryStart < 0 ? undefined : queryStart);
  if (rawPath.includes("%") || rawPath.split("/").some((part) => part === "." || part === ".."))
    return invalidLink();
  let url: URL;
  try {
    url = new URL(value);
  } catch {
    return invalidLink();
  }
  if (url.protocol !== "https:" || url.username || url.password || url.port || url.hash || url.hostname !== url.hostname.toLowerCase())
    return invalidLink();
  const target = route(url.hostname, rawPath);
  const query = parseQuery(queryStart < 0 ? "" : value.slice(queryStart + 1));
  const token = query.get("xsec_token");
  const source = query.get("xsec_source");
  for (const key of query.keys()) {
    if (SENSITIVE_KEY.test(key) && !(target.platform === "XIAOHONGSHU" && key === "xsec_token"))
      return invalidLink();
  }
  if ((token === undefined) !== (source === undefined)) return invalidLink();
  if (token !== undefined && (target.platform !== "XIAOHONGSHU" || !XSEC_TOKEN.test(token) || !XSEC_SOURCE.test(source!)))
    return invalidLink();
  const canonical_url = token === undefined
    ? target.base
    : `${target.base}?xsec_token=${encodeURIComponent(token)}&xsec_source=${encodeURIComponent(source!)}`;
  return { platform: target.platform, kind: target.kind, external_id: target.external_id, canonical_url };
}

export function planNativeCollectionLinks(
  platforms: readonly string[],
  links: readonly string[],
): NativeCollectionLink[] {
  try {
    if (!Array.isArray(platforms) || platforms.length < 1 || platforms.length > 4 ||
        platforms.some((platform) => typeof platform !== "string" || !PLATFORM_SET.has(platform)) ||
        new Set(platforms).size !== platforms.length || !Array.isArray(links) || links.length < 1 || links.length > 100 ||
        links.some((link) => typeof link !== "string"))
      throw new Error();
    const planned = links.map(parseNativeCollectionLink);
    const selected = new Set(platforms);
    if (planned.some((item) => !selected.has(item.platform)) ||
        platforms.some((platform) => !planned.some((item) => item.platform === platform)) ||
        new Set(planned.map((item) => `${item.platform}:${item.kind}:${item.external_id}`)).size !== planned.length)
      throw new Error();
    return planned;
  } catch {
    throw new Error(INVALID_SCOPE);
  }
}

File: pilot/native_collection_links.py
"""Offline native-link planning. Recognition grants no collection authority."""
from __future__ import annotations

import re
import unicodedata
from urllib.parse import quote, unquote_plus


_PLATFORMS = ('BILIBILI', 'DOUYIN', 'XIAOHONGSHU', 'ZHIHU')
_URL = re.compile(r'https://([a-z.]+)(/[^?#]*)(?:\?([^#]*))?')
_NUMBER = r'[1-9][0-9]{0,19}'
_SLUG = r'[A-Za-z0-9_-]{1,128}'
_XHS_ID = r'[0-9a-f]{24}'
_SENSITIVE = ('token', 'cookie', 'session', 'authorization', 'signature', 'password', 'secret')
# Route matching is performed on the raw path, before any URL normalization.
_ROUTES = (
    ('www.bilibili.com', rf'/video/(BV[A-Za-z0-9]{{10}})', 'BILIBILI', 'detail', ''),
    ('space.bilibili.com', rf'/({_NUMBER})', 'BILIBILI', 'creator', ''),
    ('www.douyin.com', rf'/video/({_NUMBER})', 'DOUYIN', 'detail', ''),
    ('www.douyin.com', rf'/user/({_SLUG})', 'DOUYIN', 'creator', ''),
    ('www.xiaohongshu.com', rf'/explore/({_XHS_ID})', 'XIAOHONGSHU', 'detail', ''),
    ('www.xiaohongshu.com', rf'/user/profile/({_XHS_ID})', 'XIAOHONGSHU', 'creator', ''),
    ('www.zhihu.com', rf'/question/{_NUMBER}/answer/({_NUMBER})', 'ZHIHU', 'detail', 'answer:'),
    ('www.zhihu.com', rf'/p/({_NUMBER})', 'ZHIHU', 'detail', 'article:'),
    ('zhuanlan.zhihu.com', rf'/p/({_NUMBER})', 'ZHIHU', 'detail', 'article:'),
    ('www.zhihu.com', rf'/zvideo/({_NUMBER})', 'ZHIHU', 'detail', 'zvideo:'),
    ('www.zhihu.com', rf'/people/({_SLUG})', 'ZHIHU', 'creator', ''),
)


def _invalid():
    raise ValueError('INVALID_NATIVE_COLLECTION_LINK')


def _decode_query(value: str) -> str:
    if re.search(r'%(?![0-9a-fA-F]{2})', value):
        _invalid()
    try:
        decoded = unquote_plus(value, encoding='utf-8', errors='strict')
    except UnicodeError:
        _invalid()
    if any(ch.isspace() or ch == '\\' or unicodedata.category(ch) in ('Cc', 'Cf', 'Cs', 'Zl', 'Zp')
           for ch in decoded):
        _invalid()
    return decoded


def parse_native_collection_link(value: object) -> dict:
    """Return a fixed target descriptor, or a sanitized error without input data."""
    if type(value) is not str or not 1 <= len(value) <= 2048 or any(
            not 33 <= ord(ch) <= 126 or ch == '\\' for ch in value):
        _invalid()
    match = _URL.fullmatch(value)
    if not match:
        _invalid()
    host, path, raw_query = match.groups()
    target = None
    for domain, pattern, platform, kind, prefix in _ROUTES:
        route = re.fullmatch(pattern + '/?', path) if host == domain else None
        if route:
            target = dict(platform=platform, kind=kind, external_id=prefix + route[1])
            if prefix == 'article:':
                host = 'zhuanlan.zhihu.com'
            break
    if target is None:
        _invalid()
    query = {}
    for pair in raw_query.split('&') if raw_query else ():
        if not pair or '=' not in pair:
            _invalid()
        key, parameter = (_decode_query(part) for part in pair.split('=', 1))
        lowered = key.lower()
        if not re.fullmatch(r'[A-Za-z0-9_.-]+', key) or lowered in query or lowered == 'modal_id':
            _invalid()
        if any(marker in lowered for marker in _SENSITIVE) and not (
                target['platform'] == 'XIAOHONGSHU' and key == 'xsec_token'):
            _invalid()
        # The key's exact spelling is required for the public XHS pair.
        if lowered == 'xsec_source' and (key != lowered or target['platform'] != 'XIAOHONGSHU'):
            _invalid()
        query[lowered] = parameter
    token, source = query.get('xsec_token'), query.get('xsec_source')
    canonical = f'https://{host}{path.rstrip("/")}'
    if token is not None or source is not None:
        if (token is None or source is None or not re.fullmatch(r'[A-Za-z0-9_+/=-]{1,1024}', token)
                or not re.fullmatch(r'[A-Za-z0-9_-]{1,80}', source)):
            _invalid()
        canonical += '?xsec_token=' + quote(token, safe='') + '&xsec_source=' + source
    return target | {'canonical_url': canonical}


def plan_native_collection_links(platforms: object, links: object) -> tuple[dict, ...]:
    """Plan every selected platform, rejecting duplicates rather than truncating."""
    try:
        if (type(platforms) not in (list, tuple) or not 1 <= len(platforms) <= 4
                or any(type(platform) is not str or platform not in _PLATFORMS for platform in platforms)
                or len(set(platforms)) != len(platforms)
                or type(links) not in (list, tuple) or not 1 <= len(links) <= 100):
            raise ValueError()
        targets = tuple(parse_native_collection_link(value) for value in links)
        if set(target['platform'] for target in targets) != set(platforms):
            raise ValueError()
        identities = {(target['platform'], target['kind'], target['external_id']) for target in targets}
        if len(identities) != len(targets):
            raise ValueError()
        return targets
    except ValueError:
        raise ValueError('INVALID_NATIVE_LINK_SCOPE') from None
